pkg/resourcecollector/virtualmachine.go

```diff
@@ -350,3 +353,194 @@ func getObjectInfo(resourceNames []string, namespace string, objectKind string)
+	objectMap map[storkapi.ObjectInfo]bool) (
+	[]kubevirtv1.VirtualMachine,
+	map[storkapi.ObjectInfo]bool,
+	error) {
+
+	if len(includeResources) == 0 {
+		return nil, objectMap, nil
+	}
+
+	vmList := make([]kubevirtv1.VirtualMachine, 0)
+	kv := kubevirtops.Instance()
+	for _, resourceInfo := range includeResources {
+		if resourceInfo.Kind != "VirtualMachine" {
+			logrus.Debugf("found nonVM entry for VM BackupObject type in includeResources name:%v, kind:%v, ignoring..", resourceInfo.Name, resourceInfo.Kind)
+			continue
+		}
+		vm, err := kv.GetVirtualMachine(resourceInfo.Name, resourceInfo.Namespace)
+		if err != nil {
+			// if we could not fetch the VM specified, we error out.
+			// we continue with rest of the list if the VM is not found.
+			// If none of the VMs in the includeResource is found,
+			// we will still proceed. This will cause empty resource backup.
+			if k8s_errors.IsNotFound(err) {
+				objectMap[resourceInfo] = true
+				continue
+			}
+			logrus.Errorf("error fetching VM from includeResource list %v(%v)", resourceInfo.Name, resourceInfo.Namespace)
+			return nil, objectMap, err
+		}
+		objectMap[resourceInfo] = true
+		vmList = append(vmList, *vm)
+	}
+	return vmList, objectMap, nil
+}
+
+// GetVMIncludeListFromBackup returns all the VMs from various filters,
+func GetVMIncludeListFromBackup(backup *storkapi.ApplicationBackup, objectMap map[storkapi.ObjectInfo]bool) (
+	[]kubevirtv1.VirtualMachine,
+	map[storkapi.ObjectInfo]bool,
+	error) {
+
+	// First Fetch VM List from backup.Spec.IncludeResources
+	vmIncList, objectMap, err := getVmsFromIncludeResourceList(backup.Spec.IncludeResources, objectMap)
+	if err != nil {
+		return nil, objectMap, err
+	}
+	// Second fetch VM List from Namespace and Namespace label list
+	vmNsList, err := getVmsFromNamespaceList(backup.Spec.Namespaces, objectMap)
+	if err != nil {
+		return nil, objectMap, err
+	}
+	vmIncList = append(vmIncList, vmNsList...)
+	return vmIncList, objectMap, nil
+}
+
+// GetVMIncludeResourceInfoList returns VMs and VM resources in IncludeResource format.
+// Also returns preExec and postExec rule Item with freeze/thaw rule for each of the filter VMs
+func GetVMIncludeResourceInfoList(vmList []kubevirtv1.VirtualMachine, objectMap map[storkapi.ObjectInfo]bool, skipAutoVMRule bool) (
+	[]storkapi.ObjectInfo,
+	map[storkapi.ObjectInfo]bool,
+	[]storkapi.RuleItem,
+	[]storkapi.RuleItem,
+
+) {
+
+	freezeRulesItems := make([]storkapi.RuleItem, 0)
+	unFreezeRulesItems := make([]storkapi.RuleItem, 0)
+	vmResourceInfoList := make([]storkapi.ObjectInfo, 0)
+	for _, vm := range vmList {
+		addRule := false
+		resourceInfoList := GetObjectInfoFromVMResources(vm)
+		vmInfo := storkapi.ObjectInfo{
+			GroupVersionKind: metav1.GroupVersionKind{
+				Group:   "kubevirt.io",
+				Version: "v1",
+				Kind:    "VirtualMachine",
+			},
+			Name:      vm.Name,
+			Namespace: vm.Namespace,
+		}
+		if !objectMap[vmInfo] {
+			objectMap[vmInfo] = true
```

Comment:
From `GetVMIncludeListFromBackup()`  only those ns where VM is populated and reponse of vmlist if the final list with vm<-> ns mapping. under which condition would this if case be true?

Reply:
We would have mapped VMs while processing backup.Spec.IncludeResources. But for VMs in namespaces will be differrred. So we map it here.
I have added comment for this now. Thanks.